### ISSUE TITLE
[27][28] - ktformat not formating indentation correctly

### ISSUE THREAD

**Issue by jaredsburrows** [shyiko/ktlint#285]
**Before:**
```
class TestA {
    companion object {
            private const val TAG = "TestA" // 8 spaces indent - should be 4
    }
}
```

**After:**
```
class TestA {
    companion object {
            private const val TAG = "TestA" // 8 spaces indent - should be 4
    }
}
```

**Should be:**
```
class TestA {
    companion object {
        private const val TAG = "TestA" // 4 spaces indent
    }
}
```

**Comment by shyiko** [comment]
Hi Jared.
Yeah, [IndentationRule](https://github.com/shyiko/ktlint/blob/master/ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/IndentationRule.kt) is pretty basic at the moment. All it does is check that indentation is a multiple of `indent_size`. A proper implementation shouldn't be too hard (a single scan over AST to calc expected indentation should do it) but just haven't found the time to do it (yet).
PR is (more than) welcome.

**Comment by thuytrinh** [comment]
@shyiko If I'm correct, that's why the following 8-space intents weren't formatted into 4-space intents in my case:

**Before and after formatting (code hasn't changed at all)**:

```kotlin
class AwesomeGoogleFit constructor(private val context: Context) :
        GoogleApiClient.ConnectionCallbacks,
        GoogleApiClient.OnConnectionFailedListener {

    private val client: GoogleApiClient = GoogleApiClient.Builder(context, this, this)
            .addApi(Fitness.SESSIONS_API)
            .addScope(Scope(Scopes.FITNESS_ACTIVITY_READ_WRITE))
            .build()
}
```

**Expected**:

```kotlin
class AwesomeGoogleFit constructor(private val context: Context) :
    GoogleApiClient.ConnectionCallbacks,
    GoogleApiClient.OnConnectionFailedListener {

    private val client: GoogleApiClient = GoogleApiClient.Builder(context, this, this)
        .addApi(Fitness.SESSIONS_API)
        .addScope(Scope(Scopes.FITNESS_ACTIVITY_READ_WRITE))
        .build()
}
```

However if I formatted by using Android Studio, it just worked as expected. My Android Studio has been using the code style configs which were changed by `ktlint --apply-to-idea-project`.

**Comment by thuytrinh** [comment]
@shyiko But there's one thing I can't understand is that, `ktlint` still manages to format following 8-space intents into 4 spaces:

```diff
 data class AwesomeState(
-        val a: A? = null,
-        val b: Boolean = false,
-        val c: Boolean = false,
-        val d: D? = null
+    val a: A? = null,
+    val b: Boolean = false,
+    val c: Boolean = false,
+    val d: D? = null
 )
```

Why?

**Comment by shyiko** [comment]
@iboss-ptk give it try. https://github.com/shyiko/ktlint#ast & https://plugins.jetbrains.com/plugin/227-psiviewer might be useful.

@thuytrinh parameter wrapping is handled by [ParameterListWrappingRule](https://github.com/shyiko/ktlint/blob/master/ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/ParameterListWrappingRule.kt).

**Comment by iboss-ptk** [comment]
@shyiko any idea how to determine if two indentation triggers happens to be on the same line eg.

```
val a = listOf(1,2,3)
    .map { it
        .toString()
    }
```

At `.toString()`, the indentation should be 8 spaces (default config) right? But it's in the `{}` block and it's in a dot chain. How can we determine that those triggers are on the same line?

**Comment by shyiko** [comment]
@iboss-ptk not sure I understand the question. 

`\n\s*.toString()` inside block `{}` requires +8 indent. The fact that `it` is misaligned should not change that. 

If there is anything after `{` (aside from parameters) in a multi-line block (`it` in this example) it should be moved to the next line. 

In other words,
```kotlin
val a = listOf(1,2,3)
    .map { it
        .toString()
    }
```
piped through IndentRule should produce
```kotlin
val a = listOf(1,2,3)
    .map { 
		it
        	.toString()
    }
```

Other rules can decide whether to rewrite that to

```kotlin
val a = listOf(1,2,3).map { it.toString() }
``` 
or
```kotlin
val a = listOf(1,2,3)
	.map { it.toString() }
```

**Comment by iboss-ptk** [comment]
@shyiko shouldn't it produce this?

```
val a = listOf(1,2,3)
    .map {
        it
            .toString()
    }
```

**Comment by iboss-ptk** [comment]
And similar case, if `(` and `{` appears to be on the same line, should we indent stuffs in the block once or twice?

```kotlin
addOnLayoutChangeListener(object : View.OnLayoutChangeListener {
    override fun onLayoutChange(
    )
})
```

**Comment by shyiko** [comment]
@iboss-ptk sorry, I must have had tabs in there (weren't visible on my machine). https://github.com/shyiko/ktlint/issues/285#issuecomment-429596071 updated.

> And isn't that not about indentation anymore?

You can leave "If there is anything after { (aside from parameters) in a multi-line block (it in this example) it should be moved to the next line." out of IndentRule if you want (a different rule can handle that). But I expect this can complicate things. 

> I can't think of error message about indentation in this case since it looks more like something else.

Something line "Newline missing after {" should do it.

Regarding
```kotlin
addOnLayoutChangeListener(object : View.OnLayoutChangeListener {
    override fun onLayoutChange(
    )
})
```
It's indented as expected (+4).

**Comment by shyiko** [comment]
@iboss-ptk Yep, IndentRule should match IntelliJ as closely as possible (when in doubt - just check what IntelliJ does).

**Comment by shyiko** [comment]
Moved to #338.
